SO-3666: Make namespace module assigners configurable, remove ...

AAAlinaaa opened the pull request:
...priority based assigner selection policy, update source concept
module assigner to default to configured namespace/module in case of
international concepts

AAAlinaaa committed:
SO-3666: Make namespace module assigners configurable, remove ...

...priority based assigner selection policy, update source concept
module assigner to default to configured namespace/module in case of
international concepts

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/config/SnomedCoreConfiguration.java:
```diff
@@ -101,6 +102,12 @@
 	private Set<String> reasonerExcludedModuleIds = Collections.emptySet();
 	
+	@NotNull
+	private String namespaceModuleAssigner;
```
Set this to a default value (`default`) otherwise we will need an update to hundreds of configuration files 😱.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/config/SnomedCoreConfiguration.java:
```diff
@@ -101,6 +102,12 @@
+	private String namespaceModuleAssigner;
+
+	@NotNull
+	private Set<String> internationalModuleIds = Collections.emptySet();
```
This could be set to the currently known international module id set.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SnomedNamespaceAndModuleAssigner.java:
```diff
@@ -32,6 +32,11 @@
+	/**
+	 * Returns the configuration key of the assigner
+	 */
+	String getConfigurationKey();
```
How about renaming this to `getName()`?

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SnomedNamespaceAndModuleAssignerProvider.java:
```diff
@@ -15,22 +15,35 @@
+	private final Map<String, SnomedNamespaceAndModuleAssigner> assigners;
+	
+	private SnomedNamespaceAndModuleAssignerProvider() {
+		Builder<String, SnomedNamespaceAndModuleAssigner> mapBuilder = ImmutableMap.builder();
```
`ImmutableMap.builder` throws an exception when tries to register another assigner with an already registered name. It will fail to initialize the class, which leads to `NoClassDefFoundError` and Snow Owl startup error. This registration phase should report duplicate assigners as warnings in the log and should only register the first assigner.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SourceConceptNamespaceAndModuleAssigner.java:
```diff
@@ -37,29 +38,41 @@
 public final class SourceConceptNamespaceAndModuleAssigner implements SnomedNamespaceAndModuleAssigner {
```
I would rename this assigner to something else that reflects its current functionality better.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SourceConceptNamespaceAndModuleAssigner.java:
```diff
@@ -99,4 +118,9 @@ public void clear() {
+
+	@Override
+	public String getConfigurationKey() {
+		return "sourceConcept";
```
How about `extension`?

AAAlinaaa committed:
SO-3666: Set the DefaultModuleNamespace assigner as the default ...

...namespace/module assigner in the SnomeCoreConfiguration, collect
international module ids during save, rename configurationKey getter,
ommit use of immutable maps in namespace module assigner provider,
rename SourceConceptNamespaceAndModuleAssigner

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/ExtensionNamespaceAndModuleAssigner.java:
```diff
@@ -100,9 +102,21 @@ public void collectConcreteDomainModules(final Set<String> conceptIds, final Bra
+		if (internationalModuleIds.isEmpty()) {
+			SnomedRequests.prepareSearchConcept()
+				.filterByActive(true)
+				.filterByEcl(String.format("<<%s", SnomedConstants.Concepts.IHTSDO_MAINTAINED_MODULE))
```
Missing `setLimit()` or `all()` method call.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SnomedNamespaceAndModuleAssignerProvider.java:
```diff
@@ -33,10 +32,9 @@
 		Extensions.getExtensions(NAMESPACE_ASSIGNER_EXTENSION, SnomedNamespaceAndModuleAssigner.class)
-			.forEach(assigner -> mapBuilder.put(assigner.getConfigurationKey(), assigner));
-		assigners = mapBuilder.build();
+			.forEach(assigner -> assigners.put(assigner.getName(), assigner));
```
Missing warning log message when you register an assigner with an already taken name.

AAAlinaaa committed:
SO-3666: Warn user if registering more than one assigner with the ...

... same name, remove default limit from international modules query

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SnomedNamespaceAndModuleAssignerProvider.java:
```diff
@@ -29,12 +32,18 @@
 		Extensions.getExtensions(NAMESPACE_ASSIGNER_EXTENSION, SnomedNamespaceAndModuleAssigner.class)
-			.forEach(assigner -> assigners.put(assigner.getName(), assigner));
+			.forEach(assigner -> {
+				SnomedNamespaceAndModuleAssigner previousValue = assigners.put(assigner.getName(), assigner);
```
This will keep the last assigner in the registry as opposed to the designed first.
Use `containsKey(String)` to check whether you can register the assigner and if not report it as a warning.

AAAlinaaa committed:
SO-3666: Keep first registered module/namespace assigner

AAAlinaaa committed:
SO-3666: Remove unused constant

cmark reviewed (approved):
LGTM!

cmark committed:
Merge branch '6.x' into issue/SO-3666-module_of_inferred_relationship

Conflicts:
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/id/assigner/SourceConceptNamespaceAndModuleAssigner.java

cmark committed:
SO-3666: fix compile error in ExtensionNamespaceAndModuleAssigner
